### PR TITLE
Issue 315

### PR DESCRIPTION
**Pull request by bhuvan-rk**

### CODE REVIEW

File: goldstone/core/tasks.py [hunk 1]
```diff
@@ -20,6 +20,13 @@
 from goldstone.core.models import SavedSearch, CADFEventDocType, AlertSearch, \
     Alert, EmailProducer, Producer
 from goldstone.models import es_conn
+import goldstone.cinder.utils as cinder_res
+import goldstone.glance.utils as glance_res
+import goldstone.keystone.utils as keystone_res
+import goldstone.nova.utils as nova_res
+
+# import update_nodes as update_cinder_nodes
```
Comment (jxstanford) [inline goldstone/core/tasks.py]: unnecessary comment?

File: goldstone/core/tasks.py [hunk 2]
```diff
@@ -82,15 +89,17 @@ def update_persistent_graph():
...
-    from goldstone.nova.utils import update_nodes as update_nova_nodes
-
-    update_cinder_nodes()
-    update_glance_nodes()
-    update_keystone_nodes()
-    update_nova_nodes()
+
+    graph_resources = [cinder_res, glance_res, keystone_res, nova_res]
+
+    for obj in graph_resources:
```
Comment (jxstanford) [inline goldstone/core/tasks.py]: calling all the updates each time through the loop is probably not what you want

File: goldstone/core/tasks.py [hunk 3]
```diff
@@ -82,15 +89,17 @@ def update_persistent_graph():
...
+        try:
+            cinder_res.update_nodes()
+            glance_res.update_nodes()
+            keystone_res.update_nodes()
+            nova_res.update_nodes()
+        except Exception as e:
+            logger.critical(str(e))
```
Comment (jxstanford) [inline goldstone/core/tasks.py]: why not logger.exception?

File: goldstone/core/tests.py [hunk 4]
```diff
@@ -165,6 +166,34 @@ def test_inner_send_mail_function(self):
...
+
+        # Now test the result of cinder mock returning an exception
+        update_glance_mock.return_value = []
+        self.assertTrue(update_glance_mock.call_count, 1)
+        update_cinder_mock.reset_mock()
+        update_cinder_mock.side_effect = [Exception, None]
+
+        rv = update_persistent_graph()
+
+        self.assertTrue(update_cinder_mock.called)
```
Comment (jxstanford) [inline goldstone/core/tests.py]: only need one of called or call_count

File: goldstone/core/tests.py [hunk 5]
```diff
@@ -165,6 +166,34 @@ def test_inner_send_mail_function(self):
...
+        self.assertTrue(update_cinder_mock.call_count, 4)
```
Comment (jxstanford) [inline goldstone/core/tests.py]: why looking for 4 calls?

File: goldstone/core/tests.py [hunk 6]
```diff
@@ -165,6 +166,34 @@ def test_inner_send_mail_function(self):
...
+
+        # verify that glance mock still gets called
+        # even when cinder throws an exception
+        self.assertTrue(update_glance_mock.called)
+        self.assertTrue(update_glance_mock.call_count, 2)
```
Comment (jxstanford) [inline goldstone/core/tests.py]: why looking for 2 calls?

File: goldstone/core/tests.py [hunk 7]
```diff
@@ -165,6 +166,34 @@ def test_inner_send_mail_function(self):
 class TaskTests(SimpleTestCase):
     """Test task hooks."""
 
+    @mock.patch('goldstone.core.tasks.cinder_res.update_nodes')
+    @mock.patch('goldstone.core.tasks.glance_res.update_nodes')
```
Comment (jxstanford) [inline goldstone/core/tests.py]: why only mocking these two update call and not nova, keystone, etc.?

File: goldstone/core/tasks.py (hunk 3 above)
Comment (bhuvan-rk) [inline goldstone/core/tasks.py]: logger.exception throws an exception which halts the celery task. Makes the try-catch block a moot. Hence logger.critical()

...2016-02-10T00:22:09+0000 ERROR    goldstone.core.tasks 
Traceback (most recent call last):
  File "/home/app/goldstone/core/tasks.py", line 94, in update_persistent_graph
    glance_res.update_nodes()
  File "/venv/local/lib/python2.7/site-packages/mock.py", line 955, in __call__
    return _mock_self._mock_call(*args, **kwargs)
  File "/venv/local/lib/python2.7/site-packages/mock.py", line 1015, in _mock_call
    raise result
Exception
2016-02-10T00:22:09+0000 ERROR    goldstone.core.tasks 
Traceback (most recent call last):
  File "/home/app/goldstone/core/tasks.py", line 94, in update_persistent_graph
    glance_res.update_nodes()
  File "/venv/local/lib/python2.7/site-packages/mock.py", line 955, in __call__
    return _mock_self._mock_call(*args, **kwargs)
  File "/venv/local/lib/python2.7/site-packages/mock.py", line 1013, in _mock_call
    result = next(effect)
StopIteration
2016-02-10T00:22:09+0000 ERROR    goldstone.core.tasks 
Traceback (most recent call last):
  File "/home/app/goldstone/core/tasks.py", line 94, in update_persistent_graph
    glance_res.update_nodes()
  File "/venv/local/lib/python2.7/site-packages/mock.py", line 955, in __call__
    return _mock_self._mock_call(*args, **kwargs)
  File "/venv/local/lib/python2.7/site-packages/mock.py", line 1013, in _mock_call
    result = next(effect)
StopIteration

File: goldstone/core/utils.py [hunk 8]
```diff
@@ -177,8 +177,14 @@ def process_resource_type(nodetype):
...
     # instances of the desired type, and nodes of that type in the persistent
     # resource graph.
-    actual_node_data = nodetype.clouddata()
-    persistent_nodes = nodetype.objects.all()
+    try:
+        actual_node_data = nodetype.clouddata()
+        persistent_nodes = nodetype.objects.all()
+    except Exception as e:
+        # if either of these calls throws an exception,
+        # catch it and throw an empty list for both
```
Comment (jxstanford) [inline goldstone/core/utils.py]: what about logging this exception?

File: goldstone/core/tasks.py (hunk 3 above)
Comment (jxstanford) [inline goldstone/core/tasks.py]: but the tests still pass.  do tests need refinement?

**Comment by bhuvan-rk** [comment]
Actually I spoke too soon. 

logger.exception does not halt the task but raised a backtrace of the exception on the console which I mistook as stopping the test, but it does not. 

I verified that a celery task failure actually faisl the test. So, the existing test does verify that glance, nova and keystone are getting called even when cinder throws an exception. 

I've made all the changes based on the comments in the PR. Please review.

**Pull request merged**